Validate parameters passed to job group API routes

kalikiana opened the pull request:
See also #2781 for validation in JobTemplate.

See: [poo#64075](https://progress.opensuse.org/issues/64075)

okurz commented on lib/OpenQA/WebAPI/Plugin/Helpers.pm:
```diff
@@ -498,10 +498,12 @@ sub _step_thumbnail {
     my $failed = join ', ', @{$c->validation->failed};
     my $error  = "Invalid request parameters ($failed)";
-    $c->render(text => $error, status => 400);
+    return $c->render(json => {error => $error}, status => 400) if $format eq 'json';
```
I like :)

okurz commented on t/api/10-jobgroups.t:
```diff
@@ -95,15 +95,31 @@ subtest 'list job groups' => sub() {
+    if (!$t->success) {
+        diag explain $t->tx->res->json;
+        return undef;
+    }
```
What do you think about

```perl
return diag explain $t->tx->res->json unless $t->success;
```

okurz commented on t/api/10-jobgroups.t:
```diff
@@ -131,6 +147,18 @@ subtest 'create parent group' => sub() {
+            default_keep_important_logs_in_days => '4 days'
+        }
+    )->status_is(400)->json_is(
+        '/error' => 'Invalid request parameters (default_keep_important_logs_in_days, size_limit_gb)',
```
is it ensured that the validation error messages always appear on an alphabetically ordered list of parameters?

kalikiana replied:
Yes. [Mojolicious::Validator->failed](https://mojolicious.org/perldoc/Mojolicious/Validator/Validation#failed) returns a sorted array (although it's not (yet) documented).

okurz commented:
+1

UI test failed with:

```
[08:25:33] t/ui/13-admin.t ............................ 11/? 
        #   Failed test 'size edited'
        #   at t/ui/13-admin.t line 471.
        #          got: '1000'
        #     expected: ''
        # Looks like you failed 1 test of 10.

    #   Failed test 'edit some properties'
    #   at t/ui/13-admin.t line 472.
    # Looks like you failed 1 test of 4.
[08:25:33] t/ui/13-admin.t ............................ 12/? 
#   Failed test 'job property editor'
#   at t/ui/13-admin.t line 473.
    # products: {}\nscenarios: {} # additional comment\n
[08:25:33] t/ui/13-admin.t ............................ 15/? # Looks like you failed 1 test of 15.
[08:25:33] t/ui/13-admin.t ............................ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/15 subtests
```

kalikiana commented:
> +1
> 
> UI test failed with:
> 
> ```
> [08:25:33] t/ui/13-admin.t ............................ 11/? 
>         #   Failed test 'size edited'
>         #   at t/ui/13-admin.t line 471.
>         #          got: '1000'
>         #     expected: ''
>         # Looks like you failed 1 test of 10.
> 
>     #   Failed test 'edit some properties'
>     #   at t/ui/13-admin.t line 472.
>     # Looks like you failed 1 test of 4.
> [08:25:33] t/ui/13-admin.t ............................ 12/? 
> #   Failed test 'job property editor'
> #   at t/ui/13-admin.t line 473.
>     # products: {}\nscenarios: {} # additional comment\n
> [08:25:33] t/ui/13-admin.t ............................ 15/? # Looks like you failed 1 test of 15.
> [08:25:33] t/ui/13-admin.t ............................ Dubious, test returned 1 (wstat 256, 0x100)
> Failed 1/15 subtests
> ```

After spending considerable time debugging this, my finding is this: [$self-validation->param](https://mojolicious.org/perldoc/Mojolicious/Validator/Validation#output) consistently returns no value if the value is indeed `''` as can be the case with `size_limit_gb` regardless of errors or even without any regex defined. So I changed it back to `$self->param` and added a `XXX` comment.

kraih commented:
I've proposed an upstream change that might help with this. https://github.com/mojolicious/mojo/pull/1482

Martchus commented on t/api/10-jobgroups.t:
```diff
@@ -306,31 +423,25 @@ subtest 'prevent create/update duplicate job group on top level' => sub() {
-                'The group name must not be empty or blank',
-                'Unable to create parent/job group with empty or blank name'
+        my %values = (missing => undef, invalid => '   ');
+        while (my ($error, $name) = each %values) {
```
I wouldn't introduce a use of `each` here. Not sure what the Perl-gurus have to say about it but as far as I know it is slower than a regular for loop.

perlpunk replied:
It shouldn't be slower.
But for reproducible results I would iterate over the sorted keys instead ;-)
same below.

kalikiana replied:
2842I'll change it. Also #2842 for existing uses so we can make this consistent.

okurz commented on t/api/10-jobgroups.t:
```diff
@@ -178,47 +228,117 @@ subtest 'update job group' => sub() {
+            },
+        ],
+        'list updated job group with overridden defaults'
+    );
```
Not sure if big blocks like this area easily readable and maintainable. How about just checking for some of the values? E.g.

```perl
    my $ret = $t->tx->res->json->[0];
    is $ret->{name}, 'opensuse', 'correct job group is updated';
    is $ret->{keep_important_logs_in_days}, 40, 'inherited keep_important_logs_in_days overridden';

```

kalikiana replied:
You´re right, for these ones we don´t need to check all the values!

Martchus commented on t/api/10-jobgroups.t:
```diff
@@ -130,18 +153,45 @@ subtest 'create parent group' => sub() {
+        {keep_important_logs_in_days    => '4 days'},
+        {keep_results_in_days           => '30 days'},
+        {keep_important_results_in_days => '1 days'},
+        {default_priority               => 'inherit'},
```
It would actually make sense to allow this so one can revert to the inherited value. But it looks like that is really not supported so far so I'm fine considering this invalid.

kalikiana replied:
Yes, and since we support that with job templates one could be forgiven for trying to use that. So I picked the `invalid` value here as well to highlight that it is not supported.

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1/JobGroup.pm:
```diff
@@ -192,22 +192,19 @@ sub check_top_level_group {
-=cut
-
-sub check_group_name {
+sub validate_common_properties {
```
This method should be private.

kraih commented on t/api/10-jobgroups.t:
```diff
@@ -95,15 +96,37 @@ subtest 'list job groups' => sub() {
+        {default_priority               => 'inherit'},
+        {carry_over_bugrefs             => 'no'},
+      )
+    {
```
Not critical, but i'm not really a fan of this formatting. I'd probably declare the data structure as an array before the `for` loop.
